homeassistant/components/huum/config_flow.py

```diff
@@ -0,0 +1,87 @@
+"""Config flow for huum integration."""
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+import aiohttp
+from huum.huum import Huum
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+    }
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> None:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    huum_handler = Huum(
+        data[CONF_USERNAME],
+        data[CONF_PASSWORD],
+        session=async_get_clientsession(hass),
+    )
+
+    try:
+        await huum_handler.status()
+    except aiohttp.ClientError as client_error:
+        _LOGGER.error("Could not log in to Huum with given credentials")
+        raise InvalidAuth from client_error
+
+
+class HuumConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```

Comment:
We aren't checking if the integration was already setup with the same data. Can someone has 2 saunas on one account? Can we get some kind of unique identifier?

Reply:
As far as I can tell, you only have a single sauna connected to an account, which is strange. The API does not return any unique identifiers that I can tell. So there are no unique ids for the sauna itself. So the unique identifier would then be the email used for authenticating I guess.

Reply:
> As far as I can tell, you only have a single sauna connected to an account, which is strange

when you check the data you get, do you get a list back or just an object?

I think it would be nice if we can set something unique and then also validate if this has been added before (there are quick function for that, this all can be done in 2 lines, so if you have more lines you are overcomplicating 😇 )

Reply:
Oh and when we do this, please create a test for it

Reply:
I get a single object back. So no list, no multiples, just a single entity. You can see the code in the `pyhuum` package here that calls the `/status` endpoint. https://github.com/frwickst/pyhuum/blob/main/huum/huum.py#L119

During auth there is only username and password that is passed as well, no identifier or extra value either. This would all point to there only being a single sauna that is linked to one account.

Reply:
The best possible unique identifier that I can think of in this case, is the email of the user.

Reply:
Can someone change their emailadress at huum?

Reply:
Alright, so I started to dig a bit deeper here since I also found it a bit strange that there was no identifier for the sauna. And yes, there is an identifier in the app for the sauna, but it is not accessible through the API specs that Huum provided me. So I will have to do some MiTM proxying in order to get what kind of calls that they do. This of course brings up the question regarding if this is a "good thing to do" as Huum did not provide these endpoints to me. And HA calling them could lead to thing such as them changing the APIs in the future or in the worst case blocking access for some reason. But I will at least try to get the identifier now, I'll get back to you when I've tested things.

Regarding email, it does look like the email can be changed, I've not tried it however and doing so will break the auth since it is using basic auth as their way of authentication in the API.

Reply:
So this still leads to the end result of there just being a single sauna linked to the account, but we might be able to get a unique identifier from Huum for said sauna.

Reply:
Preferably we have an identifier (so if someone resetups the integration not all automations breaks down). Otherwise we just have to check on mail address.

Reply:
Alright, so I MiTM proxied the app calls, and yes, there are things there are are used by the app to get more data, such as the sauna id and there also seem to be an endpoint to actually set the target temperature without turning on the sauna. The issue is that all of these endpoints are _not_ under the same path as the API docs that were provided by Huum. They also require a session token at times to get the value of the profile for instance, so the entire auth system is different for getting the profile data for instance.

So now the question becomes, are these things "OK" to call from Huums side. I would personally feel better getting an OK from them before starting to use them.

Reply:
If the whole auth system is different then I think it's good to exclude it for now. I think checking if the current data already exists would be enough in the config flow

Reply:
Alright, then I will look into having a unique identifier using email later on today I would say. I have emailed Huum now also and asked if they could have the profile API to be "public" as well, in order for me to use it. They are however on Christmas leave until the 1st of January, meaning that it will take some time for them to respond. We can also implement that logic later I would hope 🙂

Reply:
I mean we can do that part later. We have stuff in HA that can make migrations like these easy :)

Reply:
Ah, and it needs `self._abort_if_unique_id_configured()` as well of course to abort the setup process 😄

Reply:
Yes and no, now we know there might be a really unique id we can set later on, I'd prefer to set that as unique id when we can. For now you can use `self._async_abort_entries_match({CONF_USERNAME: user_input[CONF_USERNAME]})`

This will check the entry data of other entries if the username is taken already

Reply:
Hmmm.. doing this does not seem to stop the creation of a new entity, do I need to change `async_set_unique_id` in some other way as well. Setting it to `{CONF_USERNAME: user_input[CONF_USERNAME]}` did not seem to do it at least.

Reply:
Nevermind me, this was a test issue, just needed to pass some extra data to the mock :)

Reply:
This should now be done, hopefully 😄